demoapp/extract_mgl_sections.py

```diff
@@ -0,0 +1,319 @@
+
+    #iterate over regex list that contains both chapters and sections
+    for x in regex:  
+        items = []
+        #iterate over extracted lists
+        for item in x:
+            item = item.casefold()
+            items.append(item) 
+        if all(name in items for name in check_list):
+            for i, j in enumerate(x):
+        
+                if j.casefold() == "section":
+                    section = x[i+1].strip()
+        
+                if j.casefold() == "chapter":
+                    chapter = x[i+1].strip()
+
+                    #save current chapter in order to use it for pairing with sections mentioned later
+                    current_chapter = chapter 
+                    
+                    #add chapter and section to the list
+                    list_with_both = [chapter, section]
+                    
+                    #only keep new/unique chapter and section pairs
+                    if list_with_both not in lists_with_both:
+                        lists_with_both.append(list_with_both)
+
+        else:
+            #iterate over list that contains only sections or chapters
+            for i, j in enumerate(x):
+                #ignore SECTION with caps as it indicates sections from the bills and not the MGL
+                if j == "SECTION": 
+                    continue
+                if j == "":
+                    continue
+                
+                else:
+                    if j.casefold() == "chapter":
+                        current_chapter = x[i+1].strip() #keep track of current chapter
+                    
+                    if j.casefold() == "section":
+                        if x[i+1] == "":
+                            continue
+                        else:
+                            section = x[i+1].strip()
+                        
+                        list_with_both = [current_chapter, section]
+                        if list_with_both not in lists_with_both:
+                            lists_with_both.append(list_with_both)
+                            
+    return lists_with_both
+
+def apply_extract_sections(bills_df, bill_text_col_name, chap_sec_lists_col_name):
+    
+    """
+    Applies the 'extract_sections' function to a DataFrame containing the bills text
+
+    This function adds a new column 'Reference List' to the input DataFrame 'bills_df'. The 'DocumentText'
+    column of 'bills_df' is processed using the 'extract_sections' function, and the results are stored in
+    the newly created 'Reference List' column.
+
+    Parameters:
+    - bills_df (pd.DataFrame): The DataFrame containing bill information with a 'DocumentText' column.
+
+    Returns:
+    - bills_df (pd.DataFrame): The updated DataFrame with a new 'Reference List' column.
+    """
+
+    
+    bills_df [chap_sec_lists_col_name] = bills_df[bill_text_col_name].apply(extract_sections)
+    
+    return bills_df
+
+def make_api_call(chapter_section_list):
```

Comment:
Recommend you rename this to `query_section_text` or similar.

```suggestion
def query_section_text(chapter_section_list: tuple[str, str]) -> str | float:
```

Note that this is really looking for an input tutple (exactly two values, a string and a string) not a list (arbitrary sized container).

Note - the potential `float` output type is for the missing value case (`np.nan`).